src/Common/src/System/Security/Cryptography/DSAOpenSsl.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Cryptography;
+using Microsoft.Win32.SafeHandles;
+using System.Diagnostics;
+using System.IO;
+
+namespace System.Security.Cryptography
+{
+#if INTERNAL_ASYMMETRIC_IMPLEMENTATIONS
+    public partial class DSA : AsymmetricAlgorithm
+    {
+        public static DSA Create()
+        {
+            return new DSAImplementation.DSAOpenSsl();
+        }
+    }
+
+    internal static partial class DSAImplementation
+    {
+#endif
+        public sealed partial class DSAOpenSsl : DSA
+        {
+            private const int BitsPerByte = 8;
+            private Lazy<SafeDsaHandle> _key;
+
+            public DSAOpenSsl()
+                : this(2048)
+            {
+            }
+
+            public DSAOpenSsl(int keySize)
+            {
+                LegalKeySizesValue = s_legalKeySizes;
+                KeySize = keySize;
+                _key = new Lazy<SafeDsaHandle>(GenerateKey);
+            }
+
+            public override int KeySize
+            {
+                set
+                {
+                    if (KeySize == value)
+                    {
+                        return;
+                    }
+
+                    // Set the KeySize before FreeKey so that an invalid value doesn't throw away the key
+                    base.KeySize = value;
+
+                    FreeKey();
+                    _key = new Lazy<SafeDsaHandle>(GenerateKey);
+                }
+            }
+
+            private void ForceSetKeySize(int newKeySize)
+            {
+                // In the event that a key was loaded via ImportParameters or an IntPtr/SafeHandle
+                // it could be outside of the bounds that we currently represent as "legal key sizes".
+                // Since that is our view into the underlying component it can be detached from the
+                // component's understanding.  If it said it has opened a key, and this is the size, trust it.
+                KeySizeValue = newKeySize;
+            }
+
+            public override KeySizes[] LegalKeySizes
+            {
+                get
+                {
+                    // OpenSSL is continually being modified to accept different\higher size keys than 1024 (fips-186-3)
+                    // Some versions of OpenSSL only support fips-186-2 which limits key size to 1024.
+                    // So, copying the values from DSACryptoServiceProvider
```

Comment:
This comment doesn't really make sense.  For one, DSACryptoServiceProvider has a limit of 1024; but you seem to have gone with 3072.  Secondly; it would make more sense next to the array definition than here where it just defers to the base class...